src/libstd/process.rs

```diff
@@ -0,0 +1,835 @@
+            ExitStatus::Signal(code) =>  write!(f, "signal: {}", code),
+        }
+    }
+}
+
+impl Child {
+    /// Sends a signal to this child forcing it to exit. This is equivalent to
+    /// sending a SIGKILL on unix platforms.
+    pub fn kill(&mut self) -> IoResult<()> {
+        #[cfg(unix)] fn collect_status(p: &mut Child) {
+            // On Linux (and possibly other unices), a process that has exited will
+            // continue to accept signals because it is "defunct". The delivery of
+            // signals will only fail once the child has been reaped. For this
+            // reason, if the process hasn't exited yet, then we attempt to collect
+            // their status with WNOHANG.
+            if p.status.is_none() {
+                match p.handle.try_wait() {
+                    Some(status) => { p.status = Some(status); }
+                    None => {}
+                }
+            }
+        }
+        #[cfg(windows)] fn collect_status(_p: &mut Child) {}
+
+        collect_status(self);
+
+        // if the process has finished, and therefore had waitpid called,
+        // and we kill it, then on unix we might ending up killing a
+        // newer process that happens to have the same (re-used) id
+        if self.status.is_some() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "invalid argument: can't kill an exited process",
+                None
+            ))
+        }
+
+        unsafe { self.handle.kill(MustDieSignal) }
+    }
+
+    /// Wait for the child to exit completely, returning the status that it
+    /// exited with. This function will continue to have the same return value
+    /// after it has been called at least once.
+    ///
+    /// The stdin handle to the child process will be closed before waiting.
+    ///
+    /// # Errors
+    ///
+    /// This function can fail if a timeout was previously specified via
+    /// `set_timeout` and the timeout expires before the child exits.
+    pub fn wait(&mut self) -> IoResult<ExitStatus> {
+        drop(self.stdin.take());
+        match self.status {
+            Some(code) => Ok(code),
+            None => {
+                let status = try!(self.handle.wait(0));
+                self.status = Some(status);
+                Ok(status)
+            }
+        }
+    }
+
+    /// Simultaneously wait for the child to exit and collect all remaining
+    /// output on the stdout/stderr handles, returning a `Output`
+    /// instance.
+    ///
+    /// The stdin handle to the child is closed before waiting.
+    ///
+    /// # Errors
+    ///
+    /// This function can fail for any of the same reasons that `wait()` can
+    /// fail.
+    pub fn wait_with_output(mut self) -> IoResult<Output> {
+        drop(self.stdin.take());
+        fn read<T: Reader + Send>(stream: Option<T>) -> Receiver<old_io::IoResult<Vec<u8>>> {
+            let (tx, rx) = channel();
```

Comment:
Now that we have `JoinGuard` this may actually be able to eschew the channel entirely. I've been uncomfortable in the past about spawning threads here, but I think we do have a theoretical path forward where we don't need to spawn threads, so it may not be so bad.